mrmustard/physics/ansatze.py

```diff
@@ -0,0 +1,351 @@
+    ComplexVector,
+    Matrix,
+    Scalar,
+    Tensor,
+    Vector,
+)
+
+
+class Ansatz(ABC):
+    r"""Abstract parent class for Ansatze that we use to define quantum objects.
+    It supports all the mathematical operations (addition, subtraction, multiplication,
+    division, negation, equality, etc).
+
+    Effectively it can be thought of as a function over a continuous and/or discrete domain.
+    Note that n-dimensional arrays are like functions defined over an integer lattice of points,
+    so this class is also the parent of e.g. the Fock representation.
+    """
+
+    @abstractmethod
+    def __neg__(self) -> Ansatz:
+        ...
+
+    @abstractmethod
+    def __eq__(self, other: Ansatz) -> bool:
+        ...
+
+    @abstractmethod
+    def __add__(self, other: Ansatz) -> Ansatz:
+        ...
+
+    def __sub__(self, other: Ansatz) -> Ansatz:
+        try:
+            return self.__add__(-other)
+        except AttributeError as e:
+            raise TypeError(f"Cannot subtract {self.__class__} and {other.__class__}.") from e
+
+    @abstractmethod
+    def __call__(self, point: Any) -> Scalar:
+        r"""Evaluate the function at the given point in the domain."""
+        ...
+
+    @abstractmethod
+    def __truediv__(self, other: Union[Scalar, Ansatz]) -> Ansatz:
+        ...
+
+    @abstractmethod
+    def __mul__(self, other: Union[Scalar, Ansatz]) -> Ansatz:
+        ...
+
+    def __rmul__(self, other: Scalar) -> Ansatz:
+        return self.__mul__(other=other)
+
+
+class PolyExpBase(Ansatz):
+    r"""A family of Ansatze parametrized by a triple of a matrix, a vector and an array.
+    For example, the Bargmann representation c exp(z A z / 2 + b z) is of this form
+    (where A,b,c is the triple), or the Wigner representation (where Sigma,mu,1 is the triple).
+
+    Note that this class is not initializable (despite having an initializer)
+    because it doesn't implement all the abstract methods of Ansatz, and it is in
+    fact more general.
+    Concrete ansatze that inherit from this class will have to implement __call__,
+    __mul__ and  __matmul__, which are representation-specific.
+
+    Note that the arguments are expected to be batched, i.e. to have a batch dimension
+    or to be an iterable. This is because this class also provides the linear superposition
+    functionality by implementing the __add__ method, which concatenates the batch dimensions.
+
+    As this can blow up the number of terms in the representation, it is recommended to
+    run the `simplify()` method after adding terms together, which will combine together
+    terms that have the same exponential part.
+
+    Args:
+        mat (Batch[Matrix]):    the matrix-like data
+        vec (Batch[Vector]):    the vector-like data
+        array (Batch[Tensor]):  the array-like data
+    """
+
+    def __init__(self, mat: Batch[Matrix], vec: Batch[Vector], array: Batch[Tensor]):
+        self.mat = math.atleast_3d(math.astensor(mat))
```

Comment:
Are `mat`, `vec`, and `array` meant to be modifiable only internally to the class, or also externally (by other objects/and or users)?

If the former is true, why not replacing these with `_mat`, `_vec`, and `_array`?  You have properties that make them readable in the children classes, e.g. `mat` as `A`, `vec` as `b`, and `array` as `c`.

Reply:
I want to avoid this complication of adding private attributes and properties and getters and setters unless we have a clear reason to do it.